### PR TITLE
Improve quality guards for compression and prompts (Fixes #1054)

### LINKED ISSUE #1054: quality improvements
This was actually a prompt for a compression test, but Some great stuff in here.

Summary for packages/core/src/core/__tests__/compression-boundary.test.ts:
  ISSUES FOUND:
   1. Hardcoded values: 0.8 (compression threshold), 131134 (context limit), 0.2 (preserve threshold) - should be constants
   2. Magic numbers: Loop iterations 100, 50, 20, 10, 8, 12, 5, 15 - should be named constants
   3. Type ambiguity: vi.fn() returns unknown type - should be properly typed
   4. Hardcoded strings: 'test-runtime', 'test-provider', 'test-model', 'test-session' - should be constants
   5. Mock object creation: Empty objects {} as never used as mock settings - fragile
   6. String literals: Repeated tool names 'some_tool', 'tool-call-${i}' patterns
   7. Test data duplication: Repeated message creation patterns

  ---

  FILE 2/1546: packages/core/src/core/geminiChat.ts (2804 lines)

  ISSUES FOUND:
   1. Extremely long file: 2804 lines - should be split into multiple modules
   2. Hardcoded constants: 256 (TOKEN_SAFETY_MARGIN), 65536 (DEFAULT_COMPLETION_BUDGET), 500 (retry delay), 0.1 (temperature variation)
   3. Type ambiguity: Multiple as never casts, unknown types without proper type guards
   4. Complex nested functions: normalizeToolInteractionInput, createUserContentWithFunctionResponseFix are deeply nested
   5. Race conditions: Multiple promises (sendPromise, compressionPromise) could lead to race conditions if not properly synchronized
   6. Error handling: Generic catch blocks that swallow errors
   7. String literals: Many hardcoded strings scattered throughout
   8. Duplication: Similar retry logic in multiple methods
   9. Large methods: sendMessage, sendMessageStream, performCompression are very long
   10. Complex conditional logic: Multiple nested if/else statements without early returns
   11. Missing null checks: Some optional chaining could be more defensive
   12. TODO-style comments: Many @plan and @requirement comments suggest incomplete refactoring
   13. Hardcoded retry logic: INVALID_CONTENT_RETRY_OPTIONS values hardcoded
   14. Ambiguous variable names: result, data, value used throughout
   15. Complex state management: Multiple private fields tracking state

 I've completed analyzing 12 files so far. Here's a summary of the coding issues and bad practices found across all analyzed files:
  Summary of Issues Found (12 files analyzed)

  1. compression-boundary.test.ts
   - Magic numbers: hardcoded thresholds (0.8, 0.2, 100, 50, 131134)
   - Non-descriptive test IDs ('test-runtime', 'test-session', 'test-provider')
   - Type assertions without validation: as never, as unknown as ContentGenerator
   - Mock complexity - overly verbose test setup
   - Ambiguous test descriptions not explaining what behavior is being tested

  2. geminiChat.ts
   - Hardcoded constants: TOKEN_SAFETY_MARGIN (256), DEFAULT_COMPLETION_BUDGET (65536)
   - Magic numbers for max attempts and delays
   - Complex nested logic with unclear control flow
   - Multiple responsibility violations (compression, streaming, history management all in one class)
   - Race condition: this.cachedCompressionThreshold could be accessed during reset
   - String concatenation for error messages instead of template literals
   - Type assertions: as never, as unknown, as GenerateContentResponse
   - Ambiguous variable names: c, b in nested loops

  3. client.test.ts
   - Extensive use of vi.fn() without clear return types
   - Type assertions everywhere: as never, as unknown as
   - Magic numbers for token counts
   - Complex mock setup that's hard to maintain
   - Tests implementation details (mock interactions) rather than behavior
   - Non-descriptive variable names: mockConfigObject

  4. client.ts
   - Extremely long file (2041 lines) - single responsibility violation
   - Magic numbers: MAX_TURNS (100), complexity thresholds, token limits
   - Hardcoded TODO prompt suffix string
   - Complex nested conditionals and loops
   - Multiple responsibilities: chat management, compression, complexity analysis, todo reminders
   - Type assertions: as never, as unknown
   - Ambiguous boolean flags: forceFullIdeContext, hasChatInitialized
   - String-based type checking for tool names
   - Memory leak potential: _storedHistoryService cleared but references may persist

  5. sandwich-compression.test.ts
   - Hardcoded values in test expectations
   - Mock objects created inline without reusable factories
   - Type assertions: as never
   - Non-descriptive test IDs

  6. prompts.ts
   - Magic numbers: MAX_FOLDER_STRUCTURE_LINES (40), MAX_FOLDER_STRUCTURE_CHARS (6000)
   - Global singleton state (anti-pattern)
   - Silent error catching with console.warn
   - String manipulation for folder structure without validation
   - Tool name mapping hardcoded instead of configuration-driven

  7. createAgentRuntimeContext.ts
   - Duplicate default values for reasoning settings across files
   - String-based settings keys ('reasoning.enabled', 'reasoning.includeInContext') - typos-prone
   - Object.freeze() calls scattered throughout
   - Complex nested object creation

  8. AgentRuntimeContext.ts
   - Very long file with many interfaces
   - Inconsistent naming: ApiRequestEvent vs ApiResponseEvent vs ApiErrorEvent
   - Type aliases and interfaces mixed without clear organization
   - Optional properties used extensively (could use discriminated unions instead)

  9. compression-config.ts
   - Magic numbers hardcoded: 0.85, 0.2, 5000
   - Calculated values as constants (COMPRESSION_PRESERVE_THRESHOLD) - unclear relationship
   - No validation of threshold ranges
   - 

  10. retry.ts
   - Complex state machine with multiple counters (consecutive429s, consecutive401s)
   - Magic numbers: failoverThreshold (1), delay multipliers
   - Multiple abort signal sources that can conflict
   - Transient error detection via string matching (fragile)
   - Hardcoded error codes and phrases
   - Complex nested conditional logic
   - Side effects in retry loop (calling callbacks, modifying state)

  11. task.ts
   - Long file (~600+ lines)
   - Multiple parameter aliases (subagent_name, subagentName, subagentName) - confusing
   - Magic numbers: timeouts (60, 300)
   - Complex timeout/abort signal management
   - Race condition: multiple abort handlers that may not clean up properly
   - String-based context variables
   - Error handling with fallback messages that lose context

  12. shell.ts
   - Security: arbitrary command execution with minimal validation
   - Magic numbers: DEFAULT_SHELL_TIMEOUT_SECONDS (120), MAX_SHELL_TIMEOUT_SECONDS (600), OUTPUT_UPDATE_INTERVAL_MS (1000)
   - Complex command instrumentation logic
   - File-based IPC with temp files (security/concurrency risk)
   - Multiple abort signals (user + timeout) that can conflict
   - String parsing for command output (fragile)
   - Platform-specific code mixed throughout
   - Hardcoded tool name mapping

 ANALYZING FILE: integration-tests/ctrl-c-exit.test.ts
  Issues found:
   1. Platform-specific code smell: Windows workaround with kill() instead of testing actual Ctrl+C behavior
   2. Hardcoded timeout: 5000ms hardcoded for expecting text
   3. Accepting any exit code on Windows: expect(exitCode).not.toBeNull() is too permissive
   4. Test skip logic: Entire test skipped in CI, reducing coverage

  ---

   ANALYZING FILE: integration-tests/file-system.test.ts

  Issues found:
   1. Hardcoded test data: 'hello world' appears multiple times (should be constants)
   2. Inconsistent tool name checking: waitForAnyToolCall with hardcoded array ['write_file', 'edit', 'replace']
   3. Loose validation: validateModelOutput(result, null, ...) passes null expected content
   4. Debugging code in production tests: Multiple printDebugInfo calls clutter test logic
   5. Ambiguous error messages: Generic "Expected to find..." messages don't help debugging
   6. Platform-specific file path issues: No handling for Windows path separators

  ---

   ANALYZING FILE: packages/core/src/core/__tests__/compression-boundary.test.ts (lines 1-200)

  Issues found:
   1. Hardcoded magic numbers:
      - Compression threshold: 0.8, 0.2, 131134
      - Loop iterations: 100, 50, 20
      - Token count: 100

   2. Test implementation coupling: Directly testing private method chat['getCompressionSplit']() using bracket notation

   3. Complex test setup: Helper functions createToolCallAiMessage, createToolResponseMessage, createUserMessage - could be in a shared test utils file

   4. Ambiguous assertions: Tests like "should find a valid compression point" don't specify what "valid" means

   5. Mock complexity: Multiple nested mock objects (mockProviderAdapter, mockTelemetryAdapter, mockToolsView) - brittle and hard to maintain

   6. Type casting: } as never and } as unknown as ContentGenerator - unsafe type assertions

   7. Repeated code patterns: Multiple tests create similar history structures with slight variations - could use data-driven tests

nalysis Complete for ctrl-c-exit.test.ts & file-system.test.ts
  1. ctrl-c-exit.test.ts
  Issues Found:
   - [ERROR] Platform-specific workaround: Comment admits workaround for Windows that "forcefully kill the process, simulating a successful exit" - this is testing the workaround, not the actual behavior
   - [ERROR] Reliable test failure acknowledged: Comment states tests are "inherently flaky" in CI
   - Hardcoded timeout: 5000ms magic number without explanation
   - Exit code inconsistency: On Windows, accepts "any non-null exit code" - weak assertion

  2. file-system.test.ts  
  Issues Found:
   - [OK] Good: Uses printDebugInfo for failure debugging
   - [OK] Good: Uses validateModelOutput helper for validation
   - [ERROR] Imperative validation: Tests verify tool calls occurred but don't validate actual file behavior comprehensively
   - [ERROR] Missing edge cases: No tests for concurrent file access, permissions, large files, special characters in filenames
   - Brittle string matching: Uses toLowerCase().includes('hello') - could miss edge cases

  ---

### PR DESCRIPTION
**Pull request by acoliver**
## TLDR

Addresses the remaining issue #1054 quality items by adding compression config validation, removing unsafe casts from the compression boundary tests, and replacing the prompt-service ad-hoc singleton state with a small lifecycle manager.

## Dive Deeper

Changes included:

- Added compression config validation in packages/agents/src/core/compression-config.ts, including a module-load self-check so invalid threshold constants fail fast instead of silently shipping.
- Added behavioral coverage for compression threshold invariants and validation failures in packages/agents/src/core/__tests__/compression-config.test.ts.
- Reworked packages/agents/src/core/__tests__/compression-boundary.test.ts to use typed test doubles for ContentGenerator, RuntimeProvider, SettingsService, and provider runtime context instead of never/unknown casts.
- Introduced PromptServiceManager in packages/core/src/core/prompts.ts to encapsulate prompt service initialization state while preserving public APIs:
  - getCoreSystemPromptAsync
  - initializePromptSystem
  - drainPromptInstallerNotices
  - getCompressionPrompt
- Added behavioral tests for PromptServiceManager idempotent initialization, shared in-flight initialization, service reuse, and reset/reinitialize support.

## Reviewer Test Plan

Recommended validation:

    npm run test --workspace @vybestack/llxprt-code-agents -- compression-boundary.test.ts compression-config.test.ts
    npm run test --workspace @vybestack/llxprt-code-core -- promptServiceManager.test.ts prompts-async.test.ts prompts.coreMemory.test.ts prompts.test.ts
    npm run typecheck --workspace @vybestack/llxprt-code-agents && npm run typecheck --workspace @vybestack/llxprt-code-core
    npm run lint
    npm run build

Full local verification run:

    npm run test
    npm run lint
    npm run typecheck
    npm run format
    npm run build
    node scripts/start.js --profile-load ollamaglm51 "write me a haiku and nothing else"

Notes:

- The prescribed synthetic smoke command reached the provider call path but failed locally with an OpenAI insufficient credits API error. The alternate ollamaglm51 smoke test completed successfully.
- A deepthinker review found no blocking issues.

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  | ❓  |
| npx      | ❓  | ❓  | ❓  |
| Docker   | ❓  | ❓  | ❓  |
| Podman   | ❓  | -   | -   |
| Seatbelt | ❓  | -   | -   |

## Linked issues / bugs

Fixes #1054

### CODE REVIEW

File: packages/agents/src/core/compression-config.ts [hunk 1]
```diff
@@ -24,3 +28,111 @@ export const COMPRESSION_TOP_PRESERVE_THRESHOLD = 0.2;
...
+    errors.push(
+      `maxMessageCharsInPreserved must be a positive integer; got ${maxMessageCharsInPreserved}`,
+    );
+  }
+
+  // Only enforce the combined invariant when individual thresholds are in range,
+  // otherwise the individual errors already explain the problem.
+  if (
+    errors.length === 0 &&
+    topPreserveThreshold + preserveThreshold >= tokenThreshold
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/compression-config.ts]: Addressed in commit d9c13d6 by tracking threshold-field validity separately, so combined preserve validation still runs when unrelated fields are invalid. Added regression coverage for collecting both violations.

File: packages/core/src/core/promptServiceManager.test.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    process.env.LLXPRT_PROMPTS_DIR = tempDir;
+  });
+
+  afterAll(() => {
+    if (originalPromptsDir === undefined) {
+      delete process.env.LLXPRT_PROMPTS_DIR;
+    } else {
+      process.env.LLXPRT_PROMPTS_DIR = originalPromptsDir;
+    }
+  });
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/core/src/core/promptServiceManager.test.ts:42]: Addressed in commit d9c13d6 by removing the temporary prompts directory during afterAll cleanup.

File: packages/agents/src/core/__tests__/compression-config.test.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    const result = validateCompressionConfig({
+      tokenThreshold: 0.5,
+      preserveThreshold: 0.3,
+      topPreserveThreshold: 0.3,
+      maxMessageCharsInPreserved: -10,
+    } satisfies CompressionConfig);
+    expect(result.valid).toBe(false);
+    expect(result.errors).toHaveLength(2);
+    expect(result.errors[0]).toContain('maxMessageCharsInPreserved');
+    expect(result.errors[1]).toContain('combined preserve');
```
[the remark this answers is not in the document]
Comment (acoliver) [reply packages/agents/src/core/__tests__/compression-config.test.ts]: Addressed in commit 602b3df by making the error assertions order-independent with arrayContaining while preserving the length assertion.
